Enable UnifiedPDF compile time flag on Cocoa platforms

hortont424 opened the pull request:
#### dc6a01c9fd207e1b1394a1cb8261b410a1909cf6
<pre>
Enable UnifiedPDF compile time flag on Cocoa platforms
<a href="https://bugs.webkit.org/show_bug.cgi?id=269440">https://bugs.webkit.org/show_bug.cgi?id=269440</a>

Reviewed by Simon Fraser.

* Source/WTF/wtf/PlatformEnableCocoa.h:
Turn it on.

* Source/WTF/wtf/PlatformHave.h:
Add a define for some CG API added recently.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::pdfElementTypesForPluginPoint const):
(WebKit::UnifiedPDFPlugin::handlePDFActionForAnnotation):
Adopt defines for not-yet-everywhere API.

(WebKit::UnifiedPDFPlugin::shouldShowDebugIndicators const):
(WebKit::UnifiedPDFPlugin::didChangeSettings):
(WebKit::UnifiedPDFPlugin::paintPDFContent):
Fix the build on old compilers.

* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
Disable by default in test runner for now, so we have to explicitly opt tests into unified PDF.
We&apos;ll swap this at some point.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/WebProcess/Plugins/WebPluginInfoProvider.cpp:
(WebKit::pluginInfoVector):
(WebKit::WebPluginInfoProvider::pluginInfo):
(WebKit::WebPluginInfoProvider::webVisiblePluginInfo):
Fix a longstanding mis-assumption that if ENABLE(PDF_PLUGIN) is true, we
will always have PDF plugins (the macOS version was never runtime-disableable).
This mistake previously caused us to pretend to have a PDF plugin on iOS
when Unified PDF was enabled at build time, even if it was disabled at runtime,
breaking our fallback to PDFDocumentImage-based presentation.

Canonical link: <a href="https://commits.webkit.org/275274@main">https://commits.webkit.org/275274@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3ed164dd85b8ca8f6cd68b4fb5e3b80942abe24f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41399 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20413 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43777 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43964 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37491 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43706 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23509 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17743 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34218 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41973 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17335 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35646 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14881 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15052 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36649 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45312 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/34852 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37580 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36969 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/40719 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/41025 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16214 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/13295 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39121 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17833 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/48036 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9276 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17886 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9822 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17477 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (commented):
Ah, layout tests need rebaselining!

---

```
/Volumes/Data/worker/macOS-Monterey-Release-Build-EWS/build/Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:1447:12: error: cannot initialize return object of type 'BOOL' (aka 'signed char') with an rvalue of type 'NSURL * _Nullable'
```

Classic

----

```
In file included from /Volumes/Data/worker/macOS-Monterey-Release-Build-EWS/build/WebKitBuild/Release/DerivedSources/WebKit/unified-sources/UnifiedSource66-mm.mm:1:
In file included from /Volumes/Data/worker/macOS-Monterey-Release-Build-EWS/build/Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.mm:26:
In file included from /Volumes/Data/worker/macOS-Monterey-Release-Build-EWS/build/Source/WebKit/config.h:42:
In file included from /Volumes/Data/worker/macOS-Monterey-Release-Build-EWS/build/WebKitBuild/Release/usr/local/include/wtf/FastMalloc.h:26:
/Volumes/Data/worker/macOS-Monterey-Release-Build-EWS/build/WebKitBuild/Release/usr/local/include/wtf/StdLibExtras.h:580:5: error: static_assert failed due to requirement '!is_const<const WTF::RefPtr<WebCore::ImageBuffer, WTF::RawPtrTraits<WebCore::ImageBuffer>, WTF::DefaultRefDerefTraits<WebCore::ImageBuffer>>>::value' "T is const qualified."
    static_assert(!is_const<NonRefQualifiedType>::value, "T is const qualified.");
In file included from /Volumes/Data/worker/macOS-Monterey-Release-Build-EWS/build/WebKitBuild/Release/DerivedSources/WebKit/unified-sources/UnifiedSource66-mm.mm:1:
/Volumes/Data/worker/macOS-Monterey-Release-Build-EWS/build/Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.mm:202:46: note: in instantiation of function template specialization 'std::move<WTF::CheckMoveParameter, const WTF::RefPtr<WebCore::ImageBuffer> &>' requested here
        auto bufferAndClip = BufferAndClip { WTFMove(imageBuffer), tileRect };
In file included from /Volumes/Data/worker/macOS-Monterey-Release-Build-EWS/build/WebKitBuild/Release/DerivedSources/WebKit/unified-sources/UnifiedSource66-mm.mm:1:
In file included from /Volumes/Data/worker/macOS-Monterey-Release-Build-EWS/build/Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.mm:26:
In file included from /Volumes/Data/worker/macOS-Monterey-Release-Build-EWS/build/Source/WebKit/config.h:42:
In file included from /Volumes/Data/worker/macOS-Monterey-Release-Build-EWS/build/WebKitBuild/Release/usr/local/include/wtf/FastMalloc.h:26:
/Volumes/Data/worker/macOS-Monterey-Release-Build-EWS/build/WebKitBuild/Release/usr/local/include/wtf/StdLibExtras.h:675:29: note: expanded from macro 'WTFMove'
#define WTFMove(value) std::move<WTF::CheckMoveParameter>(value)
```

Hmm, not sure what this one's about.

smfr commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -695,6 +695,10 @@
 #define ENABLE_LEGACY_PDFKIT_PLUGIN 1
 #endif
 
+#if !defined(ENABLE_UNIFIED_PDF) && PLATFORM(COCOA) && !PLATFORM(APPLETV) && !PLATFORM(WATCHOS)
```
Seems OK (I am doing this through a local override already).

hortont424 replied:
Me too

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.mm:
```diff
@@ -181,11 +181,11 @@
 }
 
-void AsyncPDFRenderer::transferBufferToMainThread(RefPtr<ImageBuffer>&& imageBuffer, const TileForGrid& tileInfo, const FloatRect& tileRect)
+void AsyncPDFRenderer::transferBufferToMainThread(RefPtr<ImageBuffer>&& buffer, const TileForGrid& tileInfo, const FloatRect& tileRect)
```
y tho

hortont424 replied:
Chaotically trying to fix the bizarro build error, don't understand it and have never seen it locally

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1444,7 +1444,7 @@ static BOOL annotationIsExternalLink(PDFAnnotation *annotation)
         return NO;
 
-    return [annotation URL];
+    return !![annotation URL];
```
Unrelated? not clear.

aprotyas replied:
Build fix 😁

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2566,13 +2566,15 @@ static IntRect computeMarqueeSelectionRect(const WebCore::IntPoint& point1, cons
         RetainPtr currentAction = actionsForAnnotation.takeLast().get();
         performPDFAction(currentAction.get());
 
+#if HAVE(PDFKIT_WITH_NEXT_ACTIONS)
```
I guess we'd need this bit?

hortont424 replied:
Yes this bit was important

hortont424 committed:
Enable UnifiedPDF compile time flag on Cocoa platforms
https://bugs.webkit.org/show_bug.cgi?id=269440

Reviewed by Simon Fraser.

* Source/WTF/wtf/PlatformEnableCocoa.h:
Turn it on.

* Source/WTF/wtf/PlatformHave.h:
Add a define for some CG API added recently.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::pdfElementTypesForPluginPoint const):
(WebKit::UnifiedPDFPlugin::handlePDFActionForAnnotation):
Adopt defines for not-yet-everywhere API.

(WebKit::UnifiedPDFPlugin::shouldShowDebugIndicators const):
(WebKit::UnifiedPDFPlugin::didChangeSettings):
(WebKit::UnifiedPDFPlugin::paintPDFContent):
Fix the build on old compilers.

* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
Disable by default in test runner for now, so we have to explicitly opt tests into unified PDF.
We'll swap this at some point.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/WebProcess/Plugins/WebPluginInfoProvider.cpp:
(WebKit::pluginInfoVector):
(WebKit::WebPluginInfoProvider::pluginInfo):
(WebKit::WebPluginInfoProvider::webVisiblePluginInfo):
Fix a longstanding mis-assumption that if ENABLE(PDF_PLUGIN) is true, we
will always have PDF plugins (the macOS version was never runtime-disableable).
This mistake previously caused us to pretend to have a PDF plugin on iOS
when Unified PDF was enabled at build time, even if it was disabled at runtime,
breaking our fallback to PDFDocumentImage-based presentation.

Canonical link: https://commits.webkit.org/275274@main
